pkg/lib/dialogs.tsx

```diff
@@ -82,99 +82,92 @@
  *   }
  * }
  *
- * If there is a situation where you want to wait until a Dialog is closed
- * after opening, you can "await Dialogs.show(<My Dialog />)"
+*
+ * - Dialogs.show(component)
  *
- * class Example extends React.Component {
- *   static contextType = DialogsContext;
+ * Calling "Dialogs.show" will render the given component as a direct
+ * child of the inner-most enclosing "WithDialogs" component.  The
+ * component is of course intended to be a dialog, such as
+ * Patternfly's "Modal".  There is only ever one of these; a second
+ * call to "show" is considered a bug and "Dialogs.close" should be called first.
  *
- *   async function handleClick() {
- *     try {
- *       const result = await Dialogs.show(<MyDialog />);
- *       console.log(result);
- *     catch (err) {
- *     }
+ * - Dialogs.close()
  *
- *   }
+ * Calling "Dialogs.close()" will close the currently open Dialog.
  *
- *   function render() {
- *     const Dialogs = this.context;
- *     return <Button onClick={() => this.handleClick}>Open dialog</Button>;
- *   }
- * }
+ * - Dialogs.isActive()
  *
- * class MyDialog extends React.Component {
- *   static contextType = DialogsContext;
- *
- *   render() {
- *     <Button onClick={() => Dialogs.close("yes")}>Yes</Button>
- *     <Button onClick={() => Dialogs.close("no")}>No</Button>
- *   }
- * }
+ * Returns `true` if a dialog is currently being shown.
  *
+ * - Dialogs.run(component, {... props})
  *
- * - Dialogs.show(component)
+ * Shows a dialog and asyncronously waits for it to close. This creates and
+ * shows a MyDialog with the given properties, plus a special "result" property
+ * which has .resolve() and .reject() methods on it.  Calling either of those
+ * will resolve the promise returned by Dialogs.run() accordingly, closing the
+ * dialog in the process.  If the dialog is closed by a call to Dialogs.close()
+ * then Dialogs.run() throws an error.  See the example:
```

Comment:
Does it? I don't see any `throw` or call to `reject` or similar. It just sets the `dialog` state to `null`, which re-renders and thus makes the dialog go away, which seems fine. The example below also just ignores the error -- that doesn't feel like good documentation or a robust recommendation.

Closing a dialog (Cancel or the Cross icon) is not really unexpected/a error -- handling it through an exception/rejection or a `resolve(null)` is a design choice, but in both cases it has to be handled explicitly -- there's any number of things that can go wrong in a dialog (e.g. creating a VM fails due to an API error), and these normally get shown in an Alert in the dialog. This is very different from a "closed/cancelled" situation.

Note that we call `Dialogs.close` dozens of time in our code, so if you want to change that to cause an error, you have some porting work to do (I'd also advise against it for the reasons above).

Can you please add a "Fail" button to the example which calls `.reject()`, and handle the two cases?

Reply:
I had this in a previous version of the code (during a ill-advised foray into attempting to port this to a class component) but removed it.

The main idea here, though, is that dialogs shown via `.run()` get their `result` object delivered to them, and then this becomes the *only* valid way to dismiss the dialog.

`Dialogs.close()` is only valid for dialogs shown via `Dialogs.show()`, which of course is essentially everything in use today, outside of the upload code in Cockpit Files.

Reply:
I re-introduced a strict state transition enforcement mechanism (which also takes .run() into account) and the integration seem to have survived that.  Thanks for insisting on this.